ENH: use NdBSpline in RegularGridInterpolator to speed up evaluations

ev-br opened the pull request:
#### Reference issue
<!--Example: Closes gh-WXYZ.-->

supersedes and closes https://github.com/scipy/scipy/pull/7903

#### What does this implement/fix?
<!--Please explain your changes.-->

Currently, spline modes of RegularGridInterpolator  recursively construct 1D spline interpolations per dimension. This requires a python-level loop over the data points, and is very slow (cf https://github.com/scipy/scipy/issues/18010)

In this PR, we instead pre-construct an NdBSpline object and use it for evaluations. This makes evaluations faster by several orders of magnitude, at the expense of the object construction being slower (in some cases, _much_ slower, see below).

One other feature we get from NdBSpline is evaluation of derivatives. 

#### Additional information
<!--Any additional information you think is important.-->

On a high level, construction of a tensor product spline is solving a linear system for the spline `coefficients`:

```
design_matrix @ coefficients = input_data
```

There are three pieces to this PR:

1. A low-level Cython routine to **construct the design matrix**. The matrix is huge and sparse: the number of rows is the number of data points, and each row for for degree-`k` splines in `d` dimensions only has `(k+1)**d` non-zero elements.
We thus construct the matrix in the CSR format. This is relatively straightforward if somewhat tedious to keep track of non-zero b-splines and where they end up in the matrix.  
This part is subject to the curse of dimensionality: we need to loop over all data points, and for each data point we loop over `(k+1)**d` terms. 

For efficiency, this is done in Cython and we unroll the inedxing operations like `np.unravel_index` and `np.ravel_multi_index`. The game here is similar to what NdBSpline evaluations do.


2. **Solve the sparse linear system.** It turns out that the default `sparse.linalg.spsolve`, which uses a direct SuperLU solver, is extremely slow and memory hungry for high dimensions and large degrees. Using `umpack` helps somewhat but is not the final answer either. The classic answer is to use iterative solvers, of which we have several. So this PR uses `gcrotmk` by default. It works in most cases, but not all, therefore we also expose to users an interface to both change the solver and provide additional arguments (for the direct solver, reordering strategies or use umfpack; for iterative solvers: the starting point, tolerance etc).

While somewhat ugly, I don't see a way around it: the trade-off between memory, CPU time and accuracy is bound to be problem-dependent (we see it even in the test suite), the best we can do is a reasonable default + control over fine details for those who need it.

Also, why `gcrotmk` and not some other iterative algorithm? No specific preference from me, and I'd be very grateful for suggestions here.

3. **Plumb though to the RegularGridInterpolator**. In this PR, I simply duplicate the RGI methods, so that `method="cubic"` uses the new NdBSpline strategy, while `method="cubic_"` (note the traling underscore!) uses the classic recursive strategy.
This is somewhat questionable as the design, I know :-). 

An alternative is to leave the RGI be and provide a parallel helper to construct and return an NdBSpline object from data. This would parallel the 1D case, where `make_interp_spline` returns a `BSpline` instance.

One other possibility is to add the underscored methods but keep the defaults. That way, `method="cubic"` would be the classic fast to construct, slow to evaluate method, and the `method="cubic_"` would be a new opt-in for slow-to-construct, fast-to-evaluate.

Yet another possibility is to remove the "classic" recursive methods completely. This I'm not sure about because only downstream testing tells us how robust the new methods are.

-------------------------------------------

At this stage, reviewing is asked for and appreciated for all the parts here. 

I'd also ping @ilayn as our linalg resident, maybe you could weight in on the sparse linalg part?
Also I'd ping @sjev for this may improve on https://github.com/scipy/scipy/issues/18010; @thearn as the original author of https://github.com/scipy/scipy/pull/7903; @AtsushiSakai you mentioned you're a heavy RGI user?
@maxnoe @sweettyler  @denis-bz @garrettwrong @kjurka @gmgunter @braingram you reported issues with RGI refactor/updates in the 1.10 release cycle, @kmuehlbauer you were refactoring RGI for your downstream uses ---
would you folks be willing to give this PR a spin?

ev-br committed:
ENH: interpolate/ndbspline: first MVP for make_ND: full matrices, np.ravel_index/unravel_multi

ev-br committed:
MAINT: interpolate/make_nd: split py/cy, rework tests

ev-br committed:
TST: test vs RGI

ev-br committed:
DOC: some comments

ev-br committed:
ENH: interpolate/ndbspl: plumb make_nd to RGI and interpn

ev-br committed:
MAINT: tabulate unravel_index

ev-br committed:
MAINT: tabulate ravel_multi_index

ev-br committed:
TST: tweak a test

ev-br committed:
MAINT: simplify c_shape computations

ev-br committed:
MAINT: move itertools.product to python

ev-br committed:
API: add NdBSpline.design_matrix

ev-br committed:
MAINT: use umfpack if available

This seems to speed up `spsolve` a bit: locally
$ python dev.py test -t scipy/interpolate/tests/test_rgi.py  -v

goes from ~90 sec to ~60 sec. Still not great but hey.

ev-br committed:
ENH: interpolate/ndbspline: allow choosing iterative or direct sparse solvers

The direct `spsolve` is way too slow. Even with umfpack, it still is.

Thus use an iterative solver by default -- pick `gcrotmk` pretty much at random.
Allow user control for the solver itself and its additional parameters
(use_umfpack for spsolve; tolerance etc etc for the iterative solvers).

ev-br committed:
ENH: interpolate: expose derivative evaluation in RGI

ev-br committed:
BUG: interpolate: cython memview dtype

ev-br committed:
BUG: interpolate: do not fail on a method change from non-spline to spline

ev-br committed:
BUG: interpolate: work around complex r.h.s. in gcrotmk

ev-br committed:
MAINT: link to an issue in the comment

ev-br commented:
Also cross-referencing https://github.com/scipy/scipy/issues/9123 cc @oesteban

oesteban commented:
cc/ @effigies, who has been pulled into these matters lately.

peterdsharpe commented:
Just wanted to voice some support that the faster evaluations (and especially, the gradients!) are a real game-changer for surrogate modeling work. Keep up the good work, and hoping this gets reviewed and merged!

kmuehlbauer commented:
@ev-br Thanks for the ping and sorry for the delay :-) I can't say much about the NdBSpline implementation here, but I can't immediately see any problems for the RGI part. :+1:

Looking forward to use this.

ev-br committed:
BUG: interpolate/RGI: fix passing non-default solvers

Kai-Striega commented on scipy/interpolate/_rgi.py:
```diff
@@ -648,7 +688,7 @@ def interpn(points, values, xi, method="linear", bounds_error=True,
 
     # perform interpolation
-    if method in ["linear", "nearest", "slinear", "cubic", "quintic", "pchip"]:
+    if method in ["linear", "nearest", "slinear", "cubic", "quintic", "pchip", "slinear_", "cubic_", "quintic_"]:
```
Nitpick: Could we use a constant here (e.g. `_ALL_METHODS` from above?)

I'd normally use a set for this, but with such a few elements it really doesn't matter

ev-br replied:
Good catch! Done.

Kai-Striega commented on scipy/interpolate/_rgi.py:
```diff
@@ -225,12 +228,15 @@ class RegularGridInterpolator:
     # see https://github.com/JohannesBuchner/regulargrid
 
-    _SPLINE_DEGREE_MAP = {"slinear": 1, "cubic": 3, "quintic": 5, 'pchip': 3}
+    _SPLINE_DEGREE_MAP = {"slinear": 1, "cubic": 3, "quintic": 5, 'pchip': 3,
+                          "slinear_": 1, "cubic_": 3, "quintic_": 5,}
+    _SPLINE_METHODS_recurive = {"slinear_", "cubic_", "quintic_", "pchip"}
```
I'm not a fan of having the `_` names. Maybe have something a bit more descriptive to show what differentiates them from the existing methods?

j-bowhay replied:
Could we not have `cubic_legacy` or similar?

ev-br replied:
Thanks for the suggestion, done.

Kai-Striega commented on scipy/interpolate/tests/test_rgi.py:
```diff
@@ -453,6 +453,10 @@ def f(x, y):
     ])
     def test_descending_points_nd(self, method, ndims, func):
+
+        if ndims == 5 and method in {"cubic", "quintic"}:
+            pytest.skip("too slow; OOM (quintic); or nearly so (cubic)")
```
Could we mark this with an `@slow`/`@xslow` decorator?

ev-br replied:
First, I don't know how to mark one value of a parameter in a parametrized test individually. This is probably doable with some bending over backwards but I'd argue it's not needed really: 
- skipping does not leave a hole in coverage, it's covered for `method='slinear'`
- the test is simply not a good unit test: it hogs memory, throttles the CPU, and eventually OOMs for me locally.

So this test really illustrates the need for keeping the "legacy" methods which work just fine for this example.

Kai-Striega replied:
That's fair. I'll mark this as resolved.

Kai-Striega commented on scipy/interpolate/tests/test_rgi.py:
```diff
@@ -864,6 +905,9 @@ def test_complex(self, method):
         v2 = v2r + 1j*v2i
+
+        # atol = 5e-6 if method == "quintic" else 1e-15
+        # assert_allclose(v1, v2, atol=atol)
```
Is there a reason these lines are still uncommented?

ev-br replied:
Thanks! Removed.

Kai-Striega reviewed (commented):
I had a quick read through. Couple of nit-picking comments, otherwise looks good. Is there anything in particular you would like me to focus on?

thearn commented:
@ev-br I appreciate seeing this being carried forward! I originally designed this capability a few years back to aid in surrogate modelling efforts. Looking through the changes now.

ev-br committed:
MAINT: interpolate/ndbspline: address review comments, add a test

ev-br commented:
> Couple of nit-picking comments, otherwise looks good. Is there anything in particular you would like me to focus on?

Thanks Kai! I addressed your comments in 09766e6 ( + Jake's suggestion for `cubic_legacy` naming, thanks Jake!).

My main concern for this PR is the UI: 
- WDYT about options 1-5 in https://mail.python.org/archives/list/scipy-dev@python.org/thread/JWTMIX4TRVSCFJM6N45VDO3PEMOUBV7K/ ?
- `solver_args` or `**solver_args`? There is an example in 09766e6 --- currently the call signature is `RGI(...., rtol=1e-7`); an alternative is `RGI(..., solver_args={"rtol": 1e-7}`.

ev-br commented:
Here's a rough performance comparison of new and "legacy" modes. Using a variant of @Kai-Striega 's script from https://github.com/scipy/scipy/issues/7663#issuecomment-1153669263 (the scipt is below the fold), 
we compare "slinear" and "slinear_legacy" modes vs interp2d. In the dataframe below, "call / interp2d" is the ratio of call times of `RGI(...., method="slinear"` and `interp2d`, ditto for "call legacy / interp2d".

```
In [56]: pd.DataFrame({'N': ns, 'call/interp2d': call_times, 'call legacy/interp2d': cal
    ...: l_times_legacy, 'call speedup': call_speedup})
Out[56]: 
      N  call/interp2d  call legacy/interp2d  call speedup
0     2       6.719132             28.177454      4.193615
1     4       6.913495             42.144120      6.095921
2     8       4.994976            225.055938     45.056459
3    16      11.602013            753.465163     64.942621
4    32      20.021864           2143.960552    107.080965
5    64      30.942767           4234.929001    136.863293
6   128      81.235990           5347.142879     65.822339
7   256      68.443329           5994.414466     87.582158
8   512      72.170206           6915.648815     95.824152
9  1024      88.639114           6714.246109     75.748118
```

So in a nutshell, 
- RGI is still quite a bit slower slower than the specialized `interp2d`
- new methods have a consistent speedup of 1-2 orders of magnitude.


<details>

```
In [33]: import timeit
    ...:
    ...: import numpy as np
    ...: from scipy.interpolate import interp2d, RegularGridInterpolator as RGI
    ...: import matplotlib.pyplot as plt
    ...:
    ...: n_repeats = 10
    ...:
    ...: ns = []
    ...: interp2d_setup_times = []
    ...: interp2d_call_times = []
    ...: rgi_setup_times = []
    ...: rgi_call_times = []
    ...:
    ...: all_times = [interp2d_setup_times,
    ...:              interp2d_call_times,
    ...:              rgi_setup_times,
    ...:              rgi_call_times]
    ...:
    ...: all_labels = ["interp2d_setup_time",
    ...:               "interp2d_call_time",
    ...:               "rgi_setup_time",
    ...:               "rgi_call_time"]
    ...:
    ...: for n in (2 ** i for i in range(1, 11)):
    ...:     print(n)
    ...:     x = np.arange(n)
    ...:     y = np.arange(n)
    ...:     rng = np.random.default_rng(12345)
    ...:     z = rng.uniform(size=(n, n))
    ...:
    ...:     # interp2d
    ...:     f = interp2d(x, y, z, kind='linear')
    ...:     f(x, y)
    ...:     assert np.all(f(x, y) == z)
    ...:
    ...:     interp2d_setup_time = timeit.timeit("f = interp2d(x, y, z, kind='linear')",
    ...:  globals=globals(), number=n_repeats)
    ...:     interp2d_call_time = timeit.timeit("f(x, y)", globals=globals(), number=n_r
    ...: epeats)
    ...:
    ...:     # RGI
    ...:     rgi = RGI((x, y), z.T, method='slinear_legacy')   # note the transpose
    ...:     xg, yg = np.meshgrid(x, y, indexing='ij', sparse=True)
    ...:     np.testing.assert_allclose(rgi((xg, yg)), z.T, atol=1e-15)
    ...:
    ...:     rgi_setup_time = timeit.timeit("rgi = RGI((x, y), z.T, method='slinear_lega
    ...: cy')", globals=globals(), number=n_repeats)
    ...:     rgi_call_time = timeit.timeit("rgi((xg, yg))", globals=globals(), number=n_
    ...: repeats)
    ...:
    ...:     # Keep track of times, divide by n_repeats to get time per call
    ...:     ns.append(n)
    ...:     interp2d_setup_times.append(interp2d_setup_time / n_repeats)
    ...:     interp2d_call_times.append(interp2d_call_time / n_repeats)
    ...:     rgi_setup_times.append(rgi_setup_time / n_repeats)
    ...:     rgi_call_times.append(rgi_call_time / n_repeats)
    ...:
    ...: for label, time in zip(all_labels, all_times):
    ...:     plt.plot(ns, time, label=label)
    ...:
    ...: plt.title("Benchmark interp2d vs RGI")
    ...: plt.legend(loc="upper left")
    ...: plt.xlabel("n")
    ...: plt.ylabel("time (s)")
    ...: plt.show()
```

ev-br committed:
BENCH: add RGI/NdBSpline benchmark

Kai-Striega commented:
> WDYT about options 1-5 in https://mail.python.org/archives/list/scipy-dev@python.org/thread/JWTMIX4TRVSCFJM6N45VDO3PEMOUBV7K/ ?

I think the best way forward would be to use the new methods as defaults, and use the `*_legacy` names to keep the old solvers to see if anyone uses them. _Maybe_ we can revisit deprecating the old solvers if no one uses them at a later point in time.

> solver_args or **solver_args? There is an example in https://github.com/scipy/scipy/commit/09766e6047a97b35465e23e025450c2075d69fdc --- currently the call signature is RGI(...., rtol=1e-7); an alternative is RGI(..., solver_args={"rtol": 1e-7}

I'm :+1: on `RGI(..., solver_args={"rtol": 1e-7})` this makes it explicit what is passed to `RGI(...)` and what is passed to the solver.

ev-br committed:
API: interpolate/RGI: switch to RGI(..., solver_args={'rtol': 1e-6})

Instead of unpacking `**solver_args`.

ev-br commented:
> I'm 👍 on RGI(..., solver_args={"rtol": 1e-7}) this makes it explicit what is passed to RGI(...) and what is passed to the solver.

Thanks! Updated the API to this suggestion.

ev-br committed:
TST: interpolate/RGI: add test for bad solver_args

ev-br committed:
DOC: interpolate/RGI: document the solver/solver_args

[doc only]

ev-br committed:
TST: interpolate/RGI: add test of evaluation of derivatives

ev-br committed:
MAINT: interpolate/ndbspline: clean up debug leftovers

ev-br commented:
Generally, "yes, but not in this PR".

More specifically,

1. To repeat what I said on the other issue (not linking it because that one's not related), the technical work to do to allow  N-dim tensor product pchip is to figure out how to construct an NdPPoly instance from a tensor product of PPoly instances (each PchipInterpolator is one, so it has its `.c` attribute; these need to be combined into the `.c` attribute of an NdPPoly). I suspect it'll look similar to what `make_nd` tests do in this PR, you'll need several `np.outer` calls (or equivalents with broadcasting, `x[:, None] * y[None, :]` etc).

This definitely is a separate PR, and it's not on my TODO list, and it does not require any insane coding skills, so totally doable.

2. Adding an `.extend` method to `NdBSpline` and `NdPPoly` is possible, too. This should start with BSpline before NdBSpline though, and the fact that so far nobody worked on this only means nobody wanted it enough yet :-). 

Since both of these enhancement requests are out of scope for this PR, let's move them to dedicated issue (or two). Feel free to reopen / repurpose gh-19840 or open new ones, whichever way works.

As a final note, the first item sounds easier since it's likely pure python (no compiled code). If you have a prototype, however rough, we'll be able to help to whip it to shape.

ev-br committed:
Merge branch 'main' into ndbspline_make_nd

ev-br commented:
ISTM all review comments have been addressed, there are no new comments for two weeks now, and the ML thread (https://mail.python.org/archives/list/scipy-dev@python.org/thread/JWTMIX4TRVSCFJM6N45VDO3PEMOUBV7K/) is quiet since Jan 2nd. I'm tempted to interpret the quiet as a lazy consensus, unless somebody wants have more time to test drive or a review?

If past experience with tweaking RGI is any guide, most stress testing of this PR is going to happen after it lands in main --- plus maybe some more in the rc phase, and then after the release. Thus it seems to make sense to decide on this one relatively early in the release cycle.

Which release cycle is a natural question: normally I'd assume 1.13, but then if 1.13 is an out-of-band release with a shorter cycle and the main focus on adapting to NumPy 2.0 changes, we may want to avoid lumping up backwards compatible considerations and defer this PR to 1.14. 

The backcompat effect of this PR is relatively minor, but may require users to adapt:
- defaults will change (users will have a knob to roll back to the status quo ante; the change is a one-liner along the lines of  `method="cubic"` --> `method="cubic" if scipy.version < "1.13" else "cubic_legacy"`)
- time complexity / memory requirements may change in the user code
- users with large datasets may need to experiment with new knobs to get the best out of the new/changed behavior.

So I guess it's a question to @tylerjereddy, whether you want this in 1.13 or should we defer to 1.14?
(I'm planning to be around and help patching things either way).

ev-br committed:
MAINT: interpolate: make NdBSpline cython code more NumPy 2.0 robust

tylerjereddy commented on scipy/interpolate/tests/test_bsplines.py:
```diff
@@ -2164,6 +2167,156 @@ def test_readonly(self):
+        t2 = (spl.t, spl_1.t)
+        c2 = spl.c[:, None] * spl_1.c[None, :]
+
+        return t2, c2, 3
```
This function never gets used, should we delete it?

tylerjereddy commented on scipy/interpolate/_ndbspline.py:
```diff
@@ -207,3 +219,137 @@ def __call__(self, xi, *, nu=None, extrapolate=None):
+        ndim = xvals.shape[-1]
+        if len(t) != ndim:
+            raise ValueError(
+                f"Data and knots are inconsistent: len(t) = {len(t)} for "
```
fwiw, this code path is never hit by the tests

tylerjereddy commented on scipy/interpolate/_ndbspline.py:
```diff
@@ -207,3 +219,137 @@ def __call__(self, xi, *, nu=None, extrapolate=None):
+    else:
+        res, info = solver(a, b, **solver_args)
+        if info != 0:
+            raise ValueError(f"{solver = } returns {info = }.")
```
these two value errors don't get hit by tests either

tylerjereddy reviewed (commented):
I added a few small comments from 20-30 minutes of poking around.

I think I'd be "ok" with `1.13.0`, may want to scan it one more time, but probably won't find too much more.

tylerjereddy commented:
28 commits, how to handle those?

ev-br committed:
MAINT: interpolate: address review comments

ev-br commented on scipy/interpolate/_bspl.pyx:
```diff
@@ -672,3 +672,173 @@ def evaluate_ndbspline(const double[:, ::1] xi,
+    Notes
+    -----
+
+    This routine is in Cython mainly because `find_interval` and `_deBoor_D` are.
```
The comment was stale, is removed in the last commit.

Meant to mean "it could have been in pure python, but `find_interval` and `_deBoor_D routines are not exposed to Python, so this routine is also in Cython". But that comment is stale anyway, this routine should stay in Cython, or maybe move down to C at some point in the future.

ev-br commented:
Thanks Tyler for the review! I've hopefully addressed your comments in the last commit. 
Re 28 commits --- there's nothing worth preserving in this history, so can squash-merge. I can squash them manually, too; did not do it yet to keep the last commit separate to keep the review history.

tylerjereddy commented:
I spent another hour today studying this and decided to (squash) merge given the positive comments and lack of objections, and especially the fact that no old test cases are being avoided, just new test cases added. There is a performance tradeoff change, but I'm confident you'll help me address any comments that come up on that in the release.

I found a few tiny things, but they didn't seem worth blocking on:

- typo of? `Whether to extrapolate out-of bounds values of raise a `ValueError``
- plural typo maybe `every elements of the points tuple`
- a few `XXX` comments to clean up maybe?
- wondered if meant to use `_recursive` in `_SPLINE_METHODS_recurive`
- three pure Python lines seemed uncovered, 2 `nu`-related ValueErrors
and 1 uknown method value error, but being a stickler for that doesn't seem
worth it

ev-br commented:
Thank you Tyler! 
https://github.com/scipy/scipy/pull/20000 hopefully addresses your comments.
